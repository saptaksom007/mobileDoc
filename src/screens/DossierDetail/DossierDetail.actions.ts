export const ActionTypes = {
  UPDATE_RELATIVE_PHOTO_REQUESTED: 'UPDATE_RELATIVE_PHOTO_REQUESTED',
  UPDATE_RELATIVE_PHOTO_SUCCEEDED: 'UPDATE_RELATIVE_PHOTO_SUCCEEDED',
  UPDATE_RELATIVE_PHOTO_FAILED: 'UPDATE_RELATIVE_PHOTO_FAILED',
  UPDATE_RELATIVE_PHOTO_CANCELLED: 'UPDATE_RELATIVE_PHOTO_CANCELLED'
}

export default class Actions {
  static updateRelativePhotoRequested(
    pictureFrom: number,
    patientUuid: string
  ) {
    return {
      type: ActionTypes.UPDATE_RELATIVE_PHOTO_REQUESTED,
      payload: { pictureFrom, patientUuid }
    }
  }
  static updateRelativePhotoSucceeded() {
    return {
      type: ActionTypes.UPDATE_RELATIVE_PHOTO_SUCCEEDED
    }
  }
  static updateRelativePhotoCancelled() {
    return {
      type: ActionTypes.UPDATE_RELATIVE_PHOTO_CANCELLED
    }
  }
  static updateRelativePhotoFailed(error: Error) {
    return {
      type: ActionTypes.UPDATE_RELATIVE_PHOTO_FAILED,
      payload: {
        message: error.message,
        error
      }
    }
  }
}
