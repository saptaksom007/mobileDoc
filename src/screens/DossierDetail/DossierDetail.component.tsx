import React, { Component } from 'react'
import { InteractionManager, ScrollView, View } from 'react-native'
import {
  connectActionSheet,
  ActionSheetProps
} from '@expo/react-native-action-sheet'
import { Card, ListItem, Button } from 'react-native-elements'
import { compose } from 'recompose'
import i18n from 'ex-react-native-i18n'
import { fontFamily } from 'components/CustomText/CustomText.styles'
import { momentWithLocales, getCalendarOutputFormat } from 'utilities/date'

import { connect } from 'react-redux'
import withSubtitle from 'screens/enhancers/withSubtitle'
import MessageMyDoctor from 'components/MessageMyDoctor/MessageMyDoctor.component'
import { Avatar } from 'components/Avatar/Avatar.component'
import BookAppointmentLink from 'components/BookAppointmentLink/BookAppointmentLink.component'
import { Color } from 'constants/Color'

import { findConversationForHealthSubjectUuid } from 'common-docdok/lib/domain/messaging/selectors/findConversation'
import { healthrelationActions } from 'common-docdok/lib/domain/healthrelation/actions'

import Actions from './DossierDetail.actions'

import styles from './DossierDetail.styles'
import { Action } from 'redux'
import { NavigationScreenProp } from 'react-navigation'

interface Props extends ActionSheetProps {
  dispatch(action: Action): void
  navigation: NavigationScreenProp<{ params: { patientId: string } }>
  patients: { [key: string]: any }
  professionals: {
    [key: string]: { salutation: string; firstName: string; lastName: string }
  }
  conversationsArray: any[]
}

export class DossierDetail extends Component<Props> {
  componentDidMount() {
    InteractionManager.runAfterInteractions(() => {
      const {
        dispatch,
        navigation: {
          state: { params }
        }
      } = this.props
      const currentPatient = this.getCurrentPatient()
      if (currentPatient) {
        dispatch(
          healthrelationActions.loadMissingProfessionals([
            currentPatient.primaryPhysicianUuid
          ])
        )
      }
      dispatch(healthrelationActions.loadPatientRequested(params.patientId))
    })
  }

  getCurrentPatient = (): any => {
    const {
      patients,
      navigation: {
        state: { params }
      }
    } = this.props
    return patients[params.patientId]
  }

  getPhysicianName = () => {
    const patient: any = this.getCurrentPatient()
    const { professionals } = this.props
    const { primaryPhysicianUuid } = patient
    const physician = professionals[primaryPhysicianUuid]
    return physician
      ? `${physician.salutation} ${physician.firstName} ${physician.lastName}`
      : 'Loading...'
  }

  openActionSheet = () => {
    const {
      showActionSheetWithOptions,
      dispatch,
      navigation: {
        state: { params }
      }
    } = this.props

    const options = [
      i18n.t('dossierDetail.actionSheet.camera'),
      i18n.t('dossierDetail.actionSheet.library'),
      i18n.t('global.cancel')
    ]
    const cancelButtonIndex = 2
    showActionSheetWithOptions(
      {
        options,
        cancelButtonIndex
      },
      buttonIndex =>
        buttonIndex !== 2 &&
        dispatch(
          Actions.updateRelativePhotoRequested(buttonIndex, params.patientId)
        )
    )
  }

  render() {
    const { conversationsArray } = this.props
    const patient: any = this.getCurrentPatient()
    const conversation: any = conversationsArray.find(
      findConversationForHealthSubjectUuid(patient.uuid)
    )
    const name = patient && `${patient.firstName} ${patient.lastName}`
    const avatar = <Avatar name={name} url={patient && patient.avatarPicture} />
    return (
      <ScrollView contentContainerStyle={styles.container}>
        <Card fontFamily={fontFamily}>
          <ListItem
            avatar={avatar}
            roundAvatar
            title={i18n.t('dossierDetail.name')}
            subtitle={name}
            hideChevron
            fontFamily={fontFamily}
          />
          <ListItem
            title={i18n.t('dossierDetail.birthday')}
            subtitle={
              patient &&
              momentWithLocales(patient.birthdate).calendar(
                undefined,
                getCalendarOutputFormat()
              )
            }
            hideChevron
            fontFamily={fontFamily}
          />
          <ListItem
            title={i18n.t('dossierDetail.gender')}
            subtitle={patient && patient.gender}
            hideChevron
            fontFamily={fontFamily}
          />
          <ListItem
            title={i18n.t('dossierDetail.primaryPhysician')}
            subtitle={this.getPhysicianName()}
            hideChevron
            fontFamily={fontFamily}
          />
          <Button
            onPress={this.openActionSheet}
            icon={{ name: 'add-a-photo' }}
            buttonStyle={styles.button}
            backgroundColor={Color.tintColor}
            title={i18n.t('profile.card.button.updateAvatar')}
            fontFamily={fontFamily}
          />
        </Card>
        {conversation && (
          <View style={styles.bottomLinks}>
            <MessageMyDoctor
              forceConversationId={conversation.id}
              containerStyle={styles.messageMyDoctor}
            />
            <BookAppointmentLink forceConversationId={conversation.id} />
          </View>
        )}
      </ScrollView>
    )
  }
}

function select({
  healthrelationCache: { patients, professionals },
  messaging: { conversations }
}: any) {
  const conversationsArray = Object.values(conversations)
  return {
    patients,
    professionals,
    conversationsArray
  }
}

export default compose<Props, any>(
  connectActionSheet,
  connect(select),
  withSubtitle('dossier.subtitle')
)(DossierDetail)
