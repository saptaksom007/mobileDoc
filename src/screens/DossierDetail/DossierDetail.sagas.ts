import { Buffer } from 'buffer'
import * as ImagePicker from 'expo-image-picker';
import { call, put, takeLatest } from 'redux-saga/effects'
import { Actions as NavAction } from 'navigation/SagaNavigation'
import { healthrelationActions } from 'common-docdok/lib/domain/healthrelation/actions'
import {
  resizePicture,
  getExpoPickerFromPickerTypeAsync,
  getImagePickerFromIndex,
} from 'utilities/image'
import i18n from 'ex-react-native-i18n'
import { asyncAction } from 'common-docdok/lib/utils/sagaUtils'

import Actions, { ActionTypes } from './DossierDetail.actions'

function* failed(err: Error) {
  yield put(Actions.updateRelativePhotoFailed(err))
  yield put(NavAction.showLocalError(i18n.t('profile.errors.updateAvatar')))
}

function* updateRelativePhoto({ payload: { pictureFrom, patientUuid } }: any) {
  try {
    const launchPickerAsync = yield call(
      getExpoPickerFromPickerTypeAsync,
      getImagePickerFromIndex(pictureFrom),
    )
    const resultPicker = yield call(launchPickerAsync, {
      aspect: [1, 1],
      base64: true,
      quality: 0.6,
      mediaTypes: ImagePicker.MediaTypeOptions.Images,
      allowsEditing: true,
    })
    if (resultPicker.cancelled) {
      yield put(Actions.updateRelativePhotoCancelled())
    } else {
      const { base64 } = yield call(resizePicture, resultPicker)
      const winner = yield asyncAction(
        healthrelationActions.uploadPatientAvatarRequested(
          patientUuid,
          new Buffer(base64, 'base64'),
        ),
      )
      if (!winner.succeeded) {
        yield failed(winner.failed.payload.error)
      } else {
        yield put(Actions.updateRelativePhotoSucceeded())
      }
    }
  } catch (err) {
    yield failed(err)
  }
}

export default [
  takeLatest(ActionTypes.UPDATE_RELATIVE_PHOTO_REQUESTED, updateRelativePhoto),
]
