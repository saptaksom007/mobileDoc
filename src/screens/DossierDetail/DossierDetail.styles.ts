import { StyleSheet } from 'react-native'
import { Layout } from 'constants/Layout'

export default StyleSheet.create({
  container: {
    marginBottom: 15,
    height: Layout.window.height / 1.25
  },
  button: {
    marginTop: 15
  },
  bottomLinks: {
    flexDirection: 'row',
    width: Layout.window.width,
    justifyContent: 'space-between',
    alignItems: 'center',
    padding: 30,
    marginTop: 15,
    paddingTop: 0,
    bottom: 0
  },
  messageMyDoctor: {
    marginTop: 0,
    width: undefined,
    height: undefined
  }
})
