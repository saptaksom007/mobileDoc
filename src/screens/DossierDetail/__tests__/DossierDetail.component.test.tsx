import 'react-native'
import React from 'react'
import renderer from 'react-test-renderer'
import { DossierDetail } from '../DossierDetail.component'
import { NavigationScreenProp } from 'react-navigation'

it('<DossierDetail/> render correctly', () => {
  const tree = renderer
    .create(
      <DossierDetail
        showActionSheetWithOptions={() => {}}
        dispatch={() => {}}
        patients={{ key: { avatarPicture: '', primaryPhysicianUuid: 'key' } }}
        professionals={{
          123: { firstName: 'test', lastName: 'last', salutation: '' }
        }}
        navigation={
          { state: { params: { patientId: 'key' } } } as NavigationScreenProp<
            any,
            any
          >
        }
        conversationsArray={[{ id: 123 }]}
      />
    )
    .toJSON()
  expect(tree).toMatchSnapshot()
})
