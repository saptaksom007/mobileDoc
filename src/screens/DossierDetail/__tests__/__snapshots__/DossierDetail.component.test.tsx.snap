// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DossierDetail/> render correctly 1`] = `
<RCTScrollView
  contentContainerStyle={
    Object {
      "height": 1067.2,
      "marginBottom": 15,
    }
  }
>
  <View />
</RCTScrollView>
`;
